### ISSUE TITLE
GPU Conv2D differs from CPU: finite near-float32-max vs inf on ResNet conv1 output (mixed_precision float32)

### ISSUE THREAD

**Issue by hiahu329** [tensorflow/tensorflow#117844]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

2.22.0-dev20260302

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

N/A

### Python version

3.10

### Bazel version

N/A

### GCC/compiler version

N/A

### CUDA/cuDNN version

12.5.1/9.2.1

### GPU model and memory

NVIDIA GeForce RTX 3090, 24GB

### Current behavior?

Observed behavior

With the same repro_minimized.npz (input, conv_w, bn_w), the same script runs the ResNet head on GPU first, then on CPU: ZeroPadding2D → Conv2D(64, kernel 7, stride 2). conv1_out is taken before BatchNormalization.

GPU: max ≈ 3.3577712e+38, no inf
CPU: max is inf, contains inf
Log: Loaded cuDNN version 92101 (cuda_dnn.cc)

Expected behavior

For the same float32 graph, weights, and input, GPU and CPU should not disagree on conv output with one side a large finite value and the other inf, unless such differences are documented as allowed.
Here are the external files required for reproducing this issue: 

[test3.zip](https://github.com/user-attachments/files/27477876/test3.zip)

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf
tf.keras.mixed_precision.set_global_policy("float32")
d = np.load("repro_minimized.npz", allow_pickle=True)
x_input = d["input"]
conv_w = d["conv_w"]
bn_w = d["bn_w"]

class ResNetHead(tf.keras.Model):
    def __init__(self):
        super().__init__()
        self.pad1 = tf.keras.layers.ZeroPadding2D(padding=3)
        self.conv1 = tf.keras.layers.Conv2D(64, 7, strides=2, padding="valid", use_bias=False)
        self.bn1 = tf.keras.layers.BatchNormalization(axis=-1, momentum=0.9, epsilon=1e-5)
        self.relu = tf.keras.layers.ReLU()
        self.captured = {}

    def call(self, x, training=False):
        x = self.pad1(x)
        x = self.conv1(x)
        self.captured["val"] = tf.cast(x, tf.float32).numpy()
        x = self.bn1(x, training=training)
        return self.relu(x)

def run(device):
    tf.keras.backend.clear_session()
    tf.config.optimizer.set_experimental_options({"remapping": False, "layout_optimizer": False})
    with tf.device(f"/{device.upper()}:0"):
        m = ResNetHead()
        m(tf.zeros((1, 224, 224, 3), tf.float32), training=False)
        m.conv1.set_weights([conv_w])
        m.bn1.set_weights(bn_w)
        x = tf.convert_to_tensor(x_input, tf.float32)
        m(x, training=False)
        return m.captured["val"]

out_g = run("gpu")
out_c = run("cpu")
g_inf = bool(np.isinf(out_g).any())
c_inf = bool(np.isinf(out_c).any())
print("GPU conv1_out max:", np.max(out_g), "has_inf:", g_inf)
print("CPU conv1_out max:", np.max(out_c), "has_inf:", c_inf)
```

### Relevant log output

```shell
GPU conv1_out max: 3.3577712e+38 has_inf: False
CPU conv1_out max: inf has_inf: True
```

**Comment by avmohansai** [comment]
Hi @hiahu329 ,
I am unable to reproduce the specified error when tested in colab, attached [gist](https://colab.sandbox.google.com/gist/avmohansai/2e1ffd4ad09dd95b5662ef3a75807270/117844_tf_2-21-0_tf-nightly.ipynb) here.Could you please confirm if my approach is correct or am i missing anything.
Thank you.

**Comment by hiahu329** [comment]
> Hi [@hiahu329](https://github.com/hiahu329) , I am unable to reproduce the specified error when tested in colab, attached [gist](https://colab.sandbox.google.com/gist/avmohansai/2e1ffd4ad09dd95b5662ef3a75807270/117844_tf_2-21-0_tf-nightly.ipynb) here.Could you please confirm if my approach is correct or am i missing anything. Thank you.

Thanks again for trying to reproduce — on my side with TF 2.21.0 on an RTX 3090 I see asymmetric behavior (GPU conv1_out max ≈ 3.36e38, has_inf: False vs CPU … inf, has_inf: True), which can depend on the exact wheel/CUDA/cuDNN/GPU and the same repro_minimized.npz bytes, not only the 2.21.0 version label; if you used a notebook, please restart the runtime after pip install before importing TF, confirm you’re using the identical .npz from the issue, and share your full two printed lines plus Loaded cuDNN version … from the log so we can compare environments.

**Comment by avmohansai** [comment]
Hi @hiahu329 ,
Thanks for your reply.
These are my 2 liners from logs, 
`name, driver_version. Tesla T4, 580.82.07` and 
```
#define CUDNN_MAJOR 9
#define CUDNN_MINOR 8
#define CUDNN_PATCHLEVEL 0
#define CUDNN_VERSION (CUDNN_MAJOR * 10000 + CUDNN_MINOR * 100 + CUDNN_PATCHLEVEL)
```
```
GPU conv1_out max: inf has_inf: True
CPU conv1_out max: inf has_inf: True
```
Am i missing anything?
Thank you.

**Comment by zifan6699** [comment]
> Hi [@hiahu329](https://github.com/hiahu329) , Thanks for your reply. These are my 2 liners from logs, `name, driver_version. Tesla T4, 580.82.07` and
> 
> ```
> #define CUDNN_MAJOR 9
> #define CUDNN_MINOR 8
> #define CUDNN_PATCHLEVEL 0
> #define CUDNN_VERSION (CUDNN_MAJOR * 10000 + CUDNN_MINOR * 100 + CUDNN_PATCHLEVEL)
> ```
> 
> ```
> GPU conv1_out max: inf has_inf: True
> CPU conv1_out max: inf has_inf: True
> ```
> 
> Am i missing anything? Thank you.

Hi, the reason you can't reproduce the bug is that you're using different GPU hardware and cuDNN version. Both are critical for reproducing this issue.
Here's the comparison:
Your setup: GPU=Tesla T4, cuDNN=9.8.0 → GPU output=inf, CPU output=inf (no divergence, bug hidden)
My setup: GPU=RTX 3090, cuDNN=9.2.1 → GPU output=huge value(3e38), CPU output=inf (divergence, bug visible)
Reasons:
cuDNN 9.8.0 uses a new convolution kernel that overflows (same as CPU), while 9.2.1 uses a kernel without overflow.
RTX 3090 and Tesla T4 have different GPU architectures, leading to different calculation paths.
Solution:
To reproduce the bug, you need to use cuDNN 9.2.x and the same GPU architecture (RTX 3090 / Ampere) as mine. Otherwise, the comparison is not meaningful.
